src/app/routes/Router.tsx

```diff
@@ -33,10 +34,39 @@ import { LegacyAppRedirect } from './LegacyAppRedirect'
 // see https://reactrouter.com/en/main/route/lazy#multiple-routes-in-a-single-file
 // Overviews are small, and the AllOverview would load all the other overviews anyway,
 // so it's propbably better to load them all at once
+
+// function foo<T extends boolean>(returnString: T): T extends true ? string : number;
+// function foo<T extends boolean>(returnString: T): string | number {
+//   return returnString ? String(Math.random()) : Math.random();
+// }
+
+// function getStatsById(userId: string, timeWindow: TimeWindow, convertJSONOutput: true): Promise<IPlayerStats>;
+// function getStatsById(userId: string, timeWindow: TimeWindow, convertJSONOutput: false): Promise<IStatsItem[]>;
+// function getStatsById(
+//     userId: string,
+//     timeWindow: TimeWindow = TimeWindow.Alltime,
+//     convertJSONOutput: boolean = true
+//   ): Promise<IPlayerStats | IStatsItem[]>
+
+// overrides to narrow type
```

Comment:
I did not track the code change down, but I now got typescript compilation errors saying that the types need to be different based on index/non-index routes. The type definitions for these types are almost identical (https://github.com/remix-run/react-router/blob/main/packages/react-router/lib/context.ts#L19-L57). 

I've used overrides here (based on index:boolean value) to handle this. Open to other suggestions 😊

Reply:
Hmm... Normally I think overrides are only needed if the runtime-code cares about the differences, and I don't think we do.

I think we can simplify this a bit if we don't explicitly type the return-type `LazyRouteFunction`, and let typescript infer it. I'm not sure why I initially defined the explicit types (might be that it was needed, or just a stylistic choice -  but it can be valuable when eg. refactoring). 

Note that I refactored `createOverviewLazyRouteFunction` a little bit, because I noticed a bug in the logic. Eg. if component was not found, while not passing `section` we still tried to return it. 
```ts
function createOverviewLazyRouteFunction(
    componentName: string,
    section?: Section
) {
    return async () => {
        const routeComponents = await import(`../../pages/overview/`)

        if (componentName in routeComponents) {
            return {
                Component:
                    routeComponents[
                        componentName as keyof typeof routeComponents
                    ],
            }
        }
       
        if (section) {
            return {
                element: <LegacyAppRedirect section={section} />,
            }
        }
        // this shouldn't really happen
        // since we just call this function with an existing name or with section
        return {
            Component: DefaultErrorRoute,
        }
    }
}

function createSectionLazyRouteFunction(
    section: Section,
    componentFileName: string
) {
    return async () => {
        try {
            return await import(
                `../../pages/${section.namePlural}/${componentFileName}`
            )
        } catch (e) {
            // means the component is not implemented yet
            // fallback to redirect to legacy
            if (isModuleNotFoundError(e)) {
                return {
                    element: (
                        <LegacyAppRedirect
                            section={section}
                            isNew={componentFileName === 'New'}
                        />
                    ),
                }
            }
        }
    }
}
```

Reply:
Thanks! 🙏

That makes sense (and is much cleaner) with the type inference.

I've just changed the type definitions for now to keep the PR related to the dependency updates. (But also makes sense to update the function to throw an error if the component is not found)